pkg/auth/providers/azure/config.go

```diff
@@ -0,0 +1,30 @@
+package azure
+
+import v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+
+const (
+	globalAzureADGraphEndpoint = "https://graph.windows.net/"
+	globalMSGraphEndpoint      = "https://graph.microsoft.com"
+	chinaAzureADGraphEndpoint  = "https://graph.chinacloudapi.cn/"
+	chinaMSGraphEndpoint       = "https://microsoftgraph.chinacloudapi.cn"
+)
+
+func authProviderEnabled(config *v32.AzureADConfig) bool {
+	return config.Enabled && config.GraphEndpoint != ""
+}
+
+func authProviderOld(config *v32.AzureADConfig) bool {
```

Comment:
I did try to rearrange a few things to read the graph endpoint from the auth config from the database, compute a field `isGraphEndpointDeprecated`, and store it on the auth provider value, which lives in memory. And then simply read it on every login or access token request.
However, there were two difficulties:
1. When the auth provider value is being initialized, the `azuread` auth config does not yet exist in the database, so I can't initially read the value from there. Some significant refactoring will be required to change this. I could in principle write some hook that would set that variable later, say on login or other similar action. But, again, a lot of careful work might be needed here.
1. It still might be hard to keep the value in memory on the auth provider and graph config endpoint on the auth config in the database synchronized, especially since we (secretly) allow users to go back by simply editing the graph endpoint straight on the db item. Catching that and updating the value in memory is then really hard, if possible.

I wouldn't say this is all terribly inefficient. First, the fetching of the auth config has been happening all the time in the old code, there just has not been this additional string comparison of the endpoints. And this string comparison is not a particularly expensive operation.
The good thing is that we will remove this check of whether the endpoint is deprecated soon, so this won't be an issue at all.